Apply the cookiecutter to LMS

seanh opened the pull request:
Apply our pyramid-app cookiecutter to LMS. Depends on https://github.com/hypothesis/cookiecutters/pull/121

These cookiecutter PRs are by their nature kind of hard to review by looking at the diff. Some recommendations on how to test it:

* Take a look at [the new `ci.yml` passing on this branch](https://github.com/hypothesis/lms/actions/workflows/ci.yml?query=branch%3Aapply-cookiecutter)
* Run `make help` to see the available commands and try them out
* Try `make clean sure` on your machine

seanh commented:
Note to self: recompile the `requirements/*.txt` files without upgrading everything

jon-betts reviewed (commented):
This is too much at once to seriously review on a line by line basis, but I can test it as a black box.

As mentioned, I think we've upgraded all the dependencies by mistake, but even with this it seems to work for me. I ran through these, mostly just checking it completed and didn't spit out any errors:

 * `make clean`
 * `make sure` (ran twice)
 * `make dev`
 * `make devdata`
 * `make docker`
 * `make docker-run` - Seems broken to me - Missing env settings?
 * `make sql`
 * `make template` - Resulted in quite a few changes
 
I didn't do any manual testing to see the app is actually working.

seanh commented:
@jon-betts Could you re-review?

1. `make template` should be clean now that https://github.com/hypothesis/cookiecutters/pull/121 is merged
2. I've fixed the `requirements/*.txt` files. Note that there *are* some deliberate changes to these files corresponding to the changes to the `requirements/*.in` files. But I've fixed this PR to not accidentally upgrade any existing packages in the `*.txt` files
3. `make devdata docker docker-run` is indeed broken but it's broken in the same way on `main` so I'm going to opt not to fix it in this PR: it's not something that this PR has changed

jon-betts commented on requirements/checkformatting.txt:
```diff
@@ -1,71 +1,52 @@
+pip==23.1.2
+    # via pip-tools
+setuptools==67.8.0
+    # via pip-tools
```
I think this is due to a difference between how we compile things and how dependabot does. If you want to resolve this so we don't get so much chatter in PRs from dependabot you can remove the `--allow-unsafe` from the requirements building.

jon-betts commented on pyproject.toml:
```diff
@@ -1,11 +1,140 @@
+source = ["lms", "tests/unit"]
+omit = [
+    "*/lms/__main__.py",
+    "lms/pshell.py",
+    "lms/migrations/*",
+    "lms/extensions/feature_flags/views/test.py",
+    "lms/views/feature_flags_test.py",
```
Running `make template` removes the lines from `pshell.py` onwards.

seanh replied:
This is a weird one:

* `make template` does not make any changes to `pshell.py` for me even if I do `make clean template`
* But if I clone another copy of the repo, switch to this branch, and run `make template` then it _does_ make the changes to `pshell.py` :confused:

This is the relevant part of the cookiecutter's `pyproject.toml` template:

https://github.com/hypothesis/cookiecutters/blob/d5ae86cc5817e86a51f689152e0dd49c76c22103/_shared/project/pyproject.toml#L57-L65

The lines that are being removed need to go in a `.cookiecutter/includes/coverage/omit` file. And indeed there is no such `omit` file on this branch, so I would expect the lines to be getting removed by `make template`. So why are these lines not being removed when I run `make template` in my original clone of this repo?

It turns out that I do have a `cookiecutter/includes/coverage/omit` file locally but this file is ignored by one of the cookiecutter's `.gitignore` rules:

https://github.com/hypothesis/cookiecutters/blob/d5ae86cc5817e86a51f689152e0dd49c76c22103/_shared/project/.gitignore#L2

I'll open an issue in the cookiecutter's repo to fix that rule. In the meantime I've force-added the missing file on this branch (even though it's `.gitignore`'d) so `make template` is clean now

seanh replied:
https://github.com/hypothesis/cookiecutters/issues/123

jon-betts reviewed (commented):
I'm not sure if these are things you want to do anything about, but there is a conflict with the formatting of some of the requirements files and dependabot. This is due to us running `pip-compile` with `--allow-unsafe` on. If we want to resolve it it can be removed.

When running `make template` I still see some chnages in `pyproject.toml`.
